src/main/java/io/pravega/connectors/flink/FlinkPravegaReader.java

```diff
@@ -320,23 +223,244 @@ private void triggerCheckpoint(String checkpointIdentifier) throws FlinkExceptio
         checkpointTrigger.triggerCheckpoint(checkpointId);
     }
 
-    /*
-     * create reader group for the associated stream names and scope
+    // ------------------------------------------------------------------------
+    //  utility
+    // ------------------------------------------------------------------------
+
+    /**
+     * Create the {@link ReaderGroup} for the current configuration.
      */
-    private ReaderGroup createReaderGroup() {
-        Map<Stream, StreamCut> streamConfigMap = new HashMap<>();
-        for (String stream: streamNames) {
-            streamConfigMap.put(Stream.of(scopeName, stream), StreamCut.UNBOUNDED);
+    protected ReaderGroup createReaderGroup() {
+        ReaderGroupManager readerGroupManager = createReaderGroupManager();
+        readerGroupManager.createReaderGroup(this.readerGroupName, readerGroupConfig);
+        return readerGroupManager.getReaderGroup(this.readerGroupName);
```

Comment:
We should invoke readerGroupManager.close() (since closing the `connectionFactory` does not close the connection to controller)

Reply:
Yeah, good point. The reader group object needs to be closed after this merge:

https://github.com/pravega/pravega/pull/2383/files#diff-67be2af2da7593612509f695490a147bR150

Looking at the checkpoint hook, it is not clear to me where it needs to happen, though.

Reply:
I too have been worried about the cleanup issue, so I added a `close` method to Flink's hook interface in 1.5.

https://github.com/apache/flink/blob/af29172bf555558e579a92f0ee3e56ed76fd7a1a/flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/MasterTriggerRestoreHook.java#L80